### PR TITLE
using forked version of trilead-ssh2 (from org.jenkins-ci)

### PR DESCRIPTION
**Pull request by miklosbarabas**
- upgrade to support newer algorithms

## Description
CloudStack is using the trilead-ssh2 library for ssh communications, but doesn't have support for the latest MAC algorithms, which can be a blocker of adopting Cloudstack for certain companies with more strict security requirements.

The issue with the current version of trilead-ssh2 is that it is only supporting older MAC algorithms, so if one would like to add a hypervisor host to a cluster and the host runs a sshd with old algos excluded, it will be unable to add the host.

There's a forked version of trilead-ssh2 which is still being maintained.
Upgrade to this fork/version of trilead-ssh2 would provide the possibility to use newer algorithms for SSH without the need to touch actual business logic. 

Resolves https://github.com/apache/cloudstack/issues/3255

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by miklosbarabas** [comment]
@DaanHoogland 
Actually the change is done, just wanted to add a test to cover this (that's why the PR was WIP), but didn't find an easy way of doing it. Any thoughts?

**Comment by DaanHoogland** [comment]
@miklosbarabas, I suppose you could call `protected static boolean canEndTheSshConnection(int waitResultTimeoutInMs, com.trilead.ssh2.Session sess, int conditions) throws SshException, InterruptedException` in a test that should throw an `InterruptedException`. I am not entirely sure why you need this change but I'm sure either a unit test or an integration test must be possible for it.
converting it back to draft for now.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: pom.xml [hunk 1]
```diff
@@ -164,7 +164,7 @@
         <cs.servicemix.version>3.3.3_1</cs.servicemix.version>
         <cs.servlet.version>4.0.1</cs.servlet.version>
         <cs.tomcat-embed-core.version>8.5.47</cs.tomcat-embed-core.version>
-        <cs.trilead.version>1.0.0-build222</cs.trilead.version>
+        <cs.trilead.version>trilead-ssh2-build-217-jenkins-17</cs.trilead.version>
```
Comment (rhtyd) [inline pom.xml:167]: One issue it seems Jenkins forked the build-217 than latest build-222

**Comment by DaanHoogland** [comment]
@rhtyd do you consider that a blocker? (or..?)

**Comment by rhtyd** [comment]
Moving this to 4.16.0.0 milestone, any kind of major dependency change should be avoided as we're in code freeze. The workaround has been documented in the 4.15 release notes.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Looks like the original authors are not working on it, but I see this repo https://github.com/jenkinsci/trilead-ssh2

**Comment by weizhouapache** [comment]
> Looks like the original authors are not working on it, but I see this repo https://github.com/jenkinsci/trilead-ssh2

jenkinci/trilead-ssh2 looks better.

I tested both few days ago. both do not work on ubuntu 20.04. it might be related to new openssh version.

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@miklosbarabas I've updated the jar dependency to latest version, can you check and test if it works for your env?

**Review by rhtyd — APPROVED** [review]
Let's merge if it passes a round of tests

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
Comparing against baseline 4.15 health check PR, don't see any new failures. LGTM.

**Comment by RodrigoDLopez** [comment]
hey folks, after this merge i'm facing troubles to build my own branches.
somehow the maven are saying that the `repo` doesn't exist.
To fix this I need to change the `pom.xml` from utils project.
Am I doing something wrong?
this is the error i receive:

`
[ERROR] Failed to execute goal on project cloud-utils: Could not resolve dependencies for project org.apache.cloudstack:cloud-utils:jar:4.15.1.0-SNAPSHOT: Failed to collect dependencies at org.jenkins-ci:trilead-ssh2:jar:build-217-jenkins-27: Failed to read artifact descriptor for org.jenkins-ci:trilead-ssh2:jar:build-217-jenkins-27: Could not transfer artifact org.jenkins-ci:trilead-ssh2:pom:build-217-jenkins-27 from/to repo.jenkins-ci.org.releases (http://repo.jenkins-ci.org/releases/): Transfer failed for http://repo.jenkins-ci.org/releases/org/jenkins-ci/trilead-ssh2/build-217-jenkins-27/trilead-ssh2-build-217-jenkins-27.pom 308 Permanent Redirect -> [Help 1]
`

**Comment by Dajeong-Park** [comment]
Same as above, I checked the error when building in centOS 8.4 environment.

[ERROR] Failed to execute goal on project cloud-utils: Could not resolve dependencies for project org.apache.cloudstack:cloud-utils:jar:4.16.0.0-SNAPSHOT: Failed to collect dependencies at org.jenkins-ci:trilead-ssh2:jar:build-217-jenkins-27: Failed to read artifact descriptor for org.jenkins-ci:trilead-ssh2:jar:build-217-jenkins-27: Could not transfer artifact org.jenkins-ci:trilead-ssh2:pom:build-217-jenkins-27 from/to repo.jenkins-ci.org.releases (http://repo.jenkins-ci.org/releases/): Transfer failed for http://repo.jenkins-ci.org/releases/org/jenkins-ci/trilead-ssh2/build-217-jenkins-27/trilead-ssh2-build-217-jenkins-27.pom 308 Permanent Redirect -> [Help 1]

**Comment by rhtyd** [comment]
@RodrigoDLopez @Dajeong-Park are you using 4.15 or master branch? Can you try updating the jenkins repo url to use https, as in https://github.com/apache/cloudstack/commit/f1c83a0b05844fce21e1ba8af31831773ddf239e

**Comment by Dajeong-Park** [comment]
It is being used as a forked source by reflecting the merged sources in real time, and the build was successful by changing to https. Thank you.

**Comment by rhtyd** [comment]
Alright @Dajeong-Park I'll cherry-pick the fix from master and get it on 4.15.

**Comment by RodrigoDLopez** [comment]
> @RodrigoDLopez @Dajeong-Park are you using 4.15 or master branch? Can you try updating the jenkins repo url to use https, as in [f1c83a0](https://github.com/apache/cloudstack/commit/f1c83a0b05844fce21e1ba8af31831773ddf239e)

I tried one of my own branches, and then i tried with 4.15. same error into both branches. I'm using the last maven version

> It is being used as a forked source by reflecting the merged sources in real time, and the build was successful by changing to https. Thank you.

@rhtyd as @Dajeong-Park said. 
we need to change the pom.xml in order to successfully build the project.

Can I do a small PR to fix this in 4.15?

**Comment by rhtyd** [comment]
@RodrigoDLopez the issue was recently fixed in 4.15,I've backported the commit to 4.15. Pl try using latest 4.15, if still hit the issue raise a PR thnx.

**Pull request merged**